### ISSUE TITLE
[RFC] Desktop notifications

### ISSUE THREAD

**Issue by kovidgoyal** [kovidgoyal/kitty#7685]
Hi all,

I have created a new kitten *notify* to allow sending desktop notifications,
even over SSH. It supports notification icon, buttons on the notification, sounds, waiting for
notification to be closed, etc. It works just like notify-send on Linux, except
that it also works over SSH and on macOS as well.

To use simply run:

```sh
    kitten notify "Good day" Hope you have a lovely day today
```

With icon and buttons

```sh
    kitten notify --wait-for-completion --button One --button Two --icon-path /some/image.png "Good day" Hello, fellow.
```

This is available in master now and the next nightly build in an hour or so.

The underlying desktop notification protocol I invented has been greatly
expanded to enable these features, for those interested, see
https://github.com/kovidgoyal/kitty/blob/master/docs/desktop-notifications.rst

kitty also has a new setting `filter_notification` to allow for sophisticated filtering and custom actions on notifications: https://github.com/kovidgoyal/kitty/blob/master/kitty/options/definition.py#L3038

cc: @dnkl @delthas @randomstuff

**Comment by delthas** [comment]
> No, this would be a Linux only feature. And also icons are small,
arbitrary images are not. I dont really want to encourage people sending
large images over the wire for notifications.

An image would typically be 64x64 pixels (displayed at ~50x50 pixels on my Gnome setup), so pretty small.

It'd be nice, if I send both `f = senpai`, and `p = icon`, that on Linux, best-effort, the icon would be set to `senpai` and that the image would be set to the `p` icon. And where unsupported only the icon would be set from `p`.

> We could but, this is very inefficient as the terminal has no knowledge
about how often to poll,

OK makes sense. How often would a TUI library poll for events though? Should we give a non-normative hint about that in the spec? Like typically every 1s?

> That's not specified in the spec

Meaning it's implementation dependent, but kitty would still pass it? `f` to `app_name` and `t` to `category`? Typically filtering notifications is done in the notificaftion daemon based on the notification fields https://github.com/emersion/mako/blob/master/doc/mako.5.scd#criteria

**Comment by kovidgoyal** [comment]
On Thu, Aug 01, 2024 at 03:24:39AM -0700, delthas wrote:
> > No, this would be a Linux only feature. And also icons are small,
> arbitrary images are not. I dont really want to encourage people sending
> large images over the wire for notifications.
>
> An image would typically be 64x64 pixels (displayed at ~50x50 pixels on my Gnome setup), so pretty small.

You would do that, I am sure. But once the capability is out there you
wont be the only one using it.

>
> It'd be nice, if I send both `f = senpai`, and `p = icon`, that on Linux, best-effort, the icon would be set to `senpai` and that the image would be set to the `p` icon. And where unsupported only the icon would be set from `p`.

But you could'nt rely on that behavior. Indeed for the majority of users
it would'nt work. So what would be the point?

>
> > We could but, this is very inefficient as the terminal has no knowledge
> about how often to poll,
>
> OK makes sense. How often would a TUI library poll for events though? Should we give a non-normative hint about that in the spec? Like typically every 1s?

It would depend on what the notification is used for. The application
has to decide if it wants to know when it was closed and how urgently.
For example, the notify kitten is meant to be used both interactively
and in scripts so I chose a polling interval of 50ms as a compromise.

If an application just needs to do something in the background in
response to notification closing it can poll slowly. If the UI is
waiting for it it can poll fast.

>
> > That's not specified in the spec
>
> Meaning it's implementation dependent, but kitty would still pass it? `f` to `app_name` and `t` to `category`? Typically filtering notifications is done in the notificaftion daemon based on the notification fields https://github.com/emersion/mako/blob/master/doc/mako.5.scd#criteria

kitty sends f currently, I am not convinced of the utility of sending t
as none of the notification daemons I am familiar with use/expose it,
but that can be easily changed in the future.

**Comment by dnkl** [comment]
@kovidgoyal thanks for all work you put in on extending the specification! [foot](https://codeberg.org/dnkl/foot), as of today, implements everything in it.

**Comment by kovidgoyal** [comment]
On Fri, Aug 02, 2024 at 12:56:27AM -0700, Daniel Eklöf wrote:
> @kovidgoyal thanks for all work you put in on extending the specification! [foot](https://codeberg.org/dnkl/foot), as of today, implements everything in it.

Most welcome, I am glad to hear foot has implemented it.

**Comment by randomstuff** [comment]
@kovidgoyal, I guess there is typo in the query for support section:

~~~
<OSC> 99 ; i=<some identifier> : p=? ; <terminator>
~~~

Should be:

~~~
<OSC> 99 ; i=<some identifier> ; p=? ; <terminator>
~~~

**Comment by kovidgoyal** [comment]
No, that's correct, p is a metadata key identifying the type of payload. So it is separated from i by a colon not a semi-colon, just as in all other forms of the OSC 99 escape code.

**Comment by kovidgoyal** [comment]
Added support for muting the sound associated with notifications and also using xdg sound theme names on Linux. macOS sadly has no system sounds as far as I can see beyond the default beep
https://github.com/kovidgoyal/kitty/commit/bd8162fa4259114f32d9e313700329f61e6099e3

**Comment by kovidgoyal** [comment]
released in version 0.36.0
